sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -82,8 +108,165 @@ case class GpuIf(
-    trueExpr, falseExpr.columnarEval(batch))
+  override def columnarEval(batch: ColumnarBatch): Any = {
+
+    val gpuTrueExpr = trueExpr.asInstanceOf[GpuExpression]
+    val gpuFalseExpr = falseExpr.asInstanceOf[GpuExpression]
+    val colTypes = GpuColumnVector.extractTypes(batch)
+
+    withResource(GpuExpressionsUtils.columnarEvalToColumn(predicateExpr, batch)) { pred =>
+      if (isAllTrue(pred)) {
+        GpuExpressionsUtils.columnarEvalToColumn(trueExpr, batch)
+      } else if (isAllFalse(pred)) {
+        GpuExpressionsUtils.columnarEvalToColumn(falseExpr, batch)
+      } else if (gpuTrueExpr.hasSideEffects || gpuFalseExpr.hasSideEffects) {
+        conditionalWithSideEffects(batch, pred, gpuTrueExpr, gpuFalseExpr, colTypes)
+      } else {
+        withResourceIfAllowed(trueExpr.columnarEval(batch)) { trueRet =>
+          withResourceIfAllowed(falseExpr.columnarEval(batch)) { falseRet =>
+            val finalRet = (trueRet, falseRet) match {
+              case (t: GpuColumnVector, f: GpuColumnVector) =>
+                pred.getBase.ifElse(t.getBase, f.getBase)
+              case (t: GpuScalar, f: GpuColumnVector) =>
+                pred.getBase.ifElse(t.getBase, f.getBase)
+              case (t: GpuColumnVector, f: GpuScalar) =>
+                pred.getBase.ifElse(t.getBase, f.getBase)
+              case (t: GpuScalar, f: GpuScalar) =>
+                pred.getBase.ifElse(t.getBase, f.getBase)
+              case (t, f) =>
+                throw new IllegalStateException(s"Unexpected inputs" +
+                  s" ($t: ${t.getClass}, $f: ${f.getClass})")
+            }
+            GpuColumnVector.from(finalRet, dataType)
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * When computing conditional expressions on the CPU, the true and false
+   * expressions are evaluated lazily, meaning that the true expression is
+   * only evaluated for rows where the predicate is true, and the false
+   * expression is only evaluated for rows where the predicate is false.
+   * This is important in the case where the expressions can have
+   * side-effects, such as throwing exceptions for invalid inputs.
+   *
+   * This method performs lazy evaluation on the GPU by first filtering the
+   * input batch into two batches - one for rows where the predicate is true
+   * and one for rows where the predicate is false. The expressions are
+   * evaluated against these batches and then the results are combined
+   * back into a single batch using the gather algorithm.
+   */
+  private def conditionalWithSideEffects(
+      batch: ColumnarBatch,
+      pred: GpuColumnVector,
+      gpuTrueExpr: GpuExpression,
+      gpuFalseExpr: GpuExpression,
+      colTypes: Array[DataType]): GpuColumnVector = {
+
+    def batchFromCv(t: GpuColumnVector, dt: DataType): ColumnarBatch = {
+      withResource(new Table(t.getBase)) { tbl =>
+        GpuColumnVector.from(tbl, Array(dt))
+      }
```

Comment:
We do not need to create a Table then throw it away to create a batch.  The ColumnarBatch can be created from the GpuColumnVector directly (e.g.: `new ColumnarBatch(Array(t), t.getRowCount.toInt)`).  However I don't see why we're creating columnar batches when what we really want are tables.  This creates a table (and throws it away) to create a columnar batch which we then pass to the new `gather` method which in turn wishes it had a table argument instead of a columnar batch.  Why not create a table and pass that to the new `gather` method directly?

Reply:
Thanks. That certainly simplifies things.